// @airgap/beacon-transport-walletconnect/dist/esm/error.js
/**
 *  @category Error
 *  @description Error that indicates the wallet returned an invalid namespace
 */
export class InvalidReceivedSessionNamespace extends Error {
    constructor(messageWc, codeWc, type, data) {
        super();
        this.messageWc = messageWc;
        this.codeWc = codeWc;
        this.data = data;
        this.name = 'InvalidReceivedSessionNamespace';
        const baseMessage = `${codeWc}: ${messageWc}.`;
        this.message = data
            ? type === 'incomplete'
                ? ` ${baseMessage} "${data}" is missing in the session namespace.`
                : ` ${baseMessage} "${data}" is invalid.`
            : baseMessage;
    }
}
/**
 *  @category Error
 *  @description Error that indicates there is no active session
 */
export class NotConnected extends Error {
    constructor() {
        super('Not connected, no active session');
        this.name = 'NotConnected';
    }
}
/**
 *  @category Error
 *  @description Error that indicates the session is invalid
 */
export class InvalidSession extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidSession';
    }
}
/**
 *  @category Error
 *  @description Error that indicates missing required permission scopes
 */
export class MissingRequiredScope extends Error {
    constructor(requiredScopes) {
        super(`Required permission scope were not granted for "${requiredScopes}"`);
        this.requiredScopes = requiredScopes;
        this.name = 'MissingRequiredScope';
    }
}
/**
 *  @category Error
 *  @description Error that indicates the active network is not specified
 */
export class ActiveNetworkUnspecified extends Error {
    constructor() {
        super('Please specify the active network using the "setActiveNetwork" method.');
        this.name = 'ActiveNetworkUnspecified';
    }
}
/**
 *  @category Error
 *  @description Error that indicates the active account is not specified
 */
export class ActiveAccountUnspecified extends Error {
    constructor() {
        super('Please specify the active account using the "setActiveAccount" method.');
        this.name = 'ActiveAccountUnspecified';
    }
}
/**
 *  @category Error
 *  @description Error that indicates the combinaison pkh-network is not part of the active session
 */
export class InvalidNetworkOrAccount extends Error {
    constructor(network, pkh) {
        super(`No permission. The combinaison "${network}" and "${pkh}" is not part of the active session.`);
        this.network = network;
        this.pkh = pkh;
        this.name = 'InvalidNetworkOrAccount';
    }
}
