// @airgap/beacon-dapp/dist/cjs/events.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeaconEventHandler = exports.defaultEventCallbacks = exports.BeaconEvent = void 0;
var beacon_dapp_1 = require("@airgap/beacon-dapp");
var beacon_types_1 = require("@airgap/beacon-types");
var beacon_core_1 = require("@airgap/beacon-core");
var shorten_string_1 = require("./utils/shorten-string");
var beacon_ui_1 = require("@airgap/beacon-ui");
var logger = new beacon_core_1.Logger('BeaconEvents');
var SUCCESS_TIMER = 5 * 1000;
/**
 * The different events that can be emitted by the beacon-sdk
 */
var BeaconEvent;
(function (BeaconEvent) {
    BeaconEvent["PERMISSION_REQUEST_SENT"] = "PERMISSION_REQUEST_SENT";
    BeaconEvent["PERMISSION_REQUEST_SUCCESS"] = "PERMISSION_REQUEST_SUCCESS";
    BeaconEvent["PERMISSION_REQUEST_ERROR"] = "PERMISSION_REQUEST_ERROR";
    BeaconEvent["OPERATION_REQUEST_SENT"] = "OPERATION_REQUEST_SENT";
    BeaconEvent["OPERATION_REQUEST_SUCCESS"] = "OPERATION_REQUEST_SUCCESS";
    BeaconEvent["OPERATION_REQUEST_ERROR"] = "OPERATION_REQUEST_ERROR";
    BeaconEvent["SIGN_REQUEST_SENT"] = "SIGN_REQUEST_SENT";
    BeaconEvent["SIGN_REQUEST_SUCCESS"] = "SIGN_REQUEST_SUCCESS";
    BeaconEvent["SIGN_REQUEST_ERROR"] = "SIGN_REQUEST_ERROR";
    // TODO: ENCRYPTION
    // ENCRYPT_REQUEST_SENT = 'ENCRYPT_REQUEST_SENT',
    // ENCRYPT_REQUEST_SUCCESS = 'ENCRYPT_REQUEST_SUCCESS',
    // ENCRYPT_REQUEST_ERROR = 'ENCRYPT_REQUEST_ERROR',
    BeaconEvent["BROADCAST_REQUEST_SENT"] = "BROADCAST_REQUEST_SENT";
    BeaconEvent["BROADCAST_REQUEST_SUCCESS"] = "BROADCAST_REQUEST_SUCCESS";
    BeaconEvent["BROADCAST_REQUEST_ERROR"] = "BROADCAST_REQUEST_ERROR";
    BeaconEvent["ACKNOWLEDGE_RECEIVED"] = "ACKNOWLEDGE_RECEIVED";
    BeaconEvent["LOCAL_RATE_LIMIT_REACHED"] = "LOCAL_RATE_LIMIT_REACHED";
    BeaconEvent["NO_PERMISSIONS"] = "NO_PERMISSIONS";
    BeaconEvent["ACTIVE_ACCOUNT_SET"] = "ACTIVE_ACCOUNT_SET";
    BeaconEvent["ACTIVE_TRANSPORT_SET"] = "ACTIVE_TRANSPORT_SET";
    BeaconEvent["SHOW_PREPARE"] = "SHOW_PREPARE";
    BeaconEvent["HIDE_UI"] = "HIDE_UI";
    BeaconEvent["PAIR_INIT"] = "PAIR_INIT";
    BeaconEvent["PAIR_SUCCESS"] = "PAIR_SUCCESS";
    BeaconEvent["CHANNEL_CLOSED"] = "CHANNEL_CLOSED";
    BeaconEvent["INTERNAL_ERROR"] = "INTERNAL_ERROR";
    BeaconEvent["UNKNOWN"] = "UNKNOWN";
})(BeaconEvent = exports.BeaconEvent || (exports.BeaconEvent = {}));
/**
 * Show a "Request sent" toast
 */
var showSentToast = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var openWalletAction, actions, link_1;
    return __generator(this, function (_a) {
        actions = [];
        if (data.walletInfo.deeplink) {
            if (data.walletInfo.type === 'web' ||
                (data.walletInfo.type === 'mobile' && (0, beacon_ui_1.isMobile)(window)) ||
                (data.walletInfo.type === 'desktop' && !(0, beacon_ui_1.isMobile)(window))) {
                link_1 = data.walletInfo.deeplink;
                openWalletAction = function () { return __awaiter(void 0, void 0, void 0, function () {
                    var a;
                    return __generator(this, function (_a) {
                        a = document.createElement('a');
                        a.setAttribute('rel', 'noopener');
                        a.setAttribute('href', link_1);
                        a.setAttribute('target', '_blank');
                        a.dispatchEvent(new MouseEvent('click', { view: window, bubbles: true, cancelable: true }));
                        return [2 /*return*/];
                    });
                }); };
            }
        }
        actions.push({
            text: "No answer from your wallet received yet. Please make sure the wallet is open.",
            isBold: true
        });
        actions.push({
            text: 'Did you make a mistake?',
            actionText: 'Cancel Request',
            actionCallback: function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, (0, beacon_dapp_1.closeToast)()];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            }); }
        });
        actions.push({
            text: 'Wallet not receiving request?',
            actionText: 'Reset Connection',
            actionCallback: function () { return __awaiter(void 0, void 0, void 0, function () {
                var resetCallback;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, (0, beacon_dapp_1.closeToast)()
                            // eslint-disable-next-line @typescript-eslint/unbound-method
                        ];
                        case 1:
                            _a.sent();
                            resetCallback = data.extraInfo.resetCallback;
                            if (!resetCallback) return [3 /*break*/, 3];
                            logger.log('showSentToast', 'resetCallback invoked');
                            return [4 /*yield*/, resetCallback()];
                        case 2:
                            _a.sent();
                            _a.label = 3;
                        case 3: return [2 /*return*/];
                    }
                });
            }); }
        });
        (0, beacon_dapp_1.openToast)({
            body: "Request sent to\u00A0 {{wallet}}",
            walletInfo: data.walletInfo,
            state: 'loading',
            actions: actions,
            openWalletAction: openWalletAction
        }).catch(function (toastError) { return console.error(toastError); });
        return [2 /*return*/];
    });
}); };
var showAcknowledgedToast = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        (0, beacon_dapp_1.openToast)({
            body: 'Awaiting confirmation in\u00A0 {{wallet}}',
            state: 'acknowledge',
            walletInfo: data.walletInfo
        }).catch(function (toastError) { return console.error(toastError); });
        return [2 /*return*/];
    });
}); };
var showPrepare = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var text;
    return __generator(this, function (_a) {
        text = data.walletInfo
            ? "Preparing Request for\u00A0 {{wallet}}..."
            : 'Preparing Request...';
        (0, beacon_dapp_1.openToast)({
            body: text,
            state: 'prepare',
            walletInfo: data.walletInfo
        }).catch(function (toastError) { return console.error(toastError); });
        return [2 /*return*/];
    });
}); };
var hideUI = function (elements) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        if (elements) {
            if (elements.includes('alert')) {
                (0, beacon_dapp_1.closeAlerts)();
            }
            if (elements.includes('toast')) {
                (0, beacon_dapp_1.closeToast)();
            }
        }
        else {
            (0, beacon_dapp_1.closeToast)();
        }
        return [2 /*return*/];
    });
}); };
/**
 * Show a "No Permission" alert
 */
var showNoPermissionAlert = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, (0, beacon_dapp_1.openAlert)({
                    title: 'No Permission',
                    body: 'Please allow the wallet to handle this type of request.'
                })];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show an error toast
 *
 * @param beaconError The beacon error
 */
var showErrorToast = function (response, buttons) { return __awaiter(void 0, void 0, void 0, function () {
    var error, actions, err, errorMessages, hasHumandReadableError, errCode, contractErrors;
    var _a, _b, _c, _d, _e, _f;
    return __generator(this, function (_g) {
        switch (_g.label) {
            case 0:
                error = response.errorResponse.errorType
                    ? beacon_core_1.BeaconError.getError(response.errorResponse.errorType, response.errorResponse.errorData)
                    : new beacon_core_1.UnknownBeaconError();
                actions = [
                    {
                        text: error.title,
                        isBold: true
                    }
                ];
                if (response.errorResponse.errorType === beacon_types_1.BeaconErrorType.TRANSACTION_INVALID_ERROR &&
                    response.errorResponse.errorData) {
                    err = response.errorResponse.errorData;
                    errorMessages = response.errorMessages;
                    hasHumandReadableError = false;
                    if (((_a = err[0]) === null || _a === void 0 ? void 0 : _a.contract_handle) && errorMessages && (errorMessages === null || errorMessages === void 0 ? void 0 : errorMessages[err[0].contract_handle])) {
                        errCode = (_d = (_c = (_b = err[1]) === null || _b === void 0 ? void 0 : _b.with) === null || _c === void 0 ? void 0 : _c.int) !== null && _d !== void 0 ? _d : (_f = (_e = err[1]) === null || _e === void 0 ? void 0 : _e.with) === null || _f === void 0 ? void 0 : _f.string;
                        contractErrors = errorMessages === null || errorMessages === void 0 ? void 0 : errorMessages[err[0].contract_handle];
                        if (errCode && (contractErrors === null || contractErrors === void 0 ? void 0 : contractErrors[errCode])) {
                            actions.push({
                                text: contractErrors === null || contractErrors === void 0 ? void 0 : contractErrors[errCode],
                                isBold: true
                            });
                            hasHumandReadableError = true;
                        }
                    }
                    if (!hasHumandReadableError) {
                        actions.push({
                            text: error.description
                        });
                    }
                    actions.push({
                        text: '',
                        actionText: 'Show Details',
                        actionCallback: function () { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, (0, beacon_dapp_1.closeToast)()];
                                    case 1:
                                        _a.sent();
                                        return [4 /*yield*/, (0, beacon_dapp_1.openAlert)({
                                                title: error.title,
                                                // eslint-disable-next-line @typescript-eslint/unbound-method
                                                body: error.fullDescription.description,
                                                data: error.fullDescription.data,
                                                buttons: buttons
                                            })];
                                    case 2:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); }
                    });
                }
                return [4 /*yield*/, (0, beacon_dapp_1.openToast)({
                        body: "{{wallet}}\u00A0 has returned an error",
                        timer: response.errorResponse.errorType === beacon_types_1.BeaconErrorType.ABORTED_ERROR
                            ? SUCCESS_TIMER
                            : undefined,
                        state: 'finished',
                        walletInfo: response.walletInfo,
                        actions: actions
                    })];
            case 1:
                _g.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show a rate limit reached toast
 */
var showRateLimitReached = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        (0, beacon_dapp_1.openAlert)({
            title: 'Error',
            body: 'Rate limit reached. Please slow down',
            buttons: [{ text: 'Done', style: 'outline' }],
            timer: 3000
        }).catch(function (toastError) { return console.error(toastError); });
        return [2 /*return*/];
    });
}); };
/**
 * Show a "connection successful" alert for 1.5 seconds
 */
var showExtensionConnectedAlert = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, (0, beacon_dapp_1.closeAlerts)()];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show a "channel closed" alert for 1.5 seconds
 */
var showChannelClosedAlert = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, (0, beacon_dapp_1.openAlert)({
                    title: 'Channel closed',
                    body: "Your peer has closed the connection.",
                    buttons: [{ text: 'Done', style: 'outline' }],
                    timer: 1500
                })];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
var showInternalErrorAlert = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var buttons, alertConfig;
    var _a;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                buttons = __spreadArray([], ((_a = data.buttons) !== null && _a !== void 0 ? _a : []), true);
                buttons.push({ text: 'Done', style: 'outline' });
                alertConfig = {
                    title: 'Internal Error',
                    body: data.text,
                    buttons: buttons
                };
                return [4 /*yield*/, (0, beacon_dapp_1.openAlert)(alertConfig)];
            case 1:
                _b.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show a connect alert with QR code
 *
 * @param data The data that is emitted by the PAIR_INIT event
 */
var showPairAlert = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var alertConfig;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                console.log('showPairAlert');
                alertConfig = {
                    title: 'Choose your preferred wallet',
                    body: "<p></p>",
                    pairingPayload: {
                        p2pSyncCode: data.p2pPeerInfo,
                        walletConnectSyncCode: data.walletConnectPeerInfo,
                        postmessageSyncCode: data.postmessagePeerInfo,
                        networkType: data.networkType
                    },
                    // eslint-disable-next-line @typescript-eslint/unbound-method
                    closeButtonCallback: data.abortedHandler,
                    disclaimerText: data.disclaimerText,
                    analytics: data.analytics,
                    featuredWallets: data.featuredWallets
                };
                return [4 /*yield*/, (0, beacon_dapp_1.openAlert)(alertConfig)];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show a "Permission Granted" alert
 *
 * @param data The data that is emitted by the PERMISSION_REQUEST_SUCCESS event
 */
var showPermissionSuccessAlert = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var output;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                output = data.output;
                return [4 /*yield*/, (0, beacon_dapp_1.openToast)({
                        body: "{{wallet}}\u00A0 has granted permission",
                        timer: SUCCESS_TIMER,
                        walletInfo: data.walletInfo,
                        state: 'finished',
                        actions: [
                            {
                                text: 'Address',
                                actionText: (0, shorten_string_1.shortenString)(output.address),
                                isBold: true
                            },
                            {
                                text: 'Network',
                                actionText: "".concat(output.network.type)
                            },
                            {
                                text: 'Permissions',
                                actionText: output.scopes.join(', ')
                            }
                        ]
                    })];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show an "Operation Broadcasted" alert
 *
 * @param data The data that is emitted by the OPERATION_REQUEST_SUCCESS event
 */
var showOperationSuccessAlert = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var account, output, blockExplorer;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                account = data.account, output = data.output, blockExplorer = data.blockExplorer;
                return [4 /*yield*/, (0, beacon_dapp_1.openToast)({
                        body: "{{wallet}}\u00A0 successfully submitted operation",
                        timer: SUCCESS_TIMER,
                        state: 'finished',
                        walletInfo: data.walletInfo,
                        actions: [
                            {
                                text: (0, shorten_string_1.shortenString)(output.transactionHash),
                                isBold: true,
                                actionText: "Open Blockexplorer",
                                actionLogo: 'external',
                                actionCallback: function () { return __awaiter(void 0, void 0, void 0, function () {
                                    var link;
                                    return __generator(this, function (_a) {
                                        switch (_a.label) {
                                            case 0: return [4 /*yield*/, blockExplorer.getTransactionLink(output.transactionHash, account.network)];
                                            case 1:
                                                link = _a.sent();
                                                window.open(link, '_blank', 'noopener');
                                                return [4 /*yield*/, (0, beacon_dapp_1.closeToast)()];
                                            case 2:
                                                _a.sent();
                                                return [2 /*return*/];
                                        }
                                    });
                                }); }
                            }
                        ]
                    })];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show a "Transaction Signed" alert
 *
 * @param data The data that is emitted by the SIGN_REQUEST_SUCCESS event
 */
var showSignSuccessAlert = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var output;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                output = data.output;
                return [4 /*yield*/, (0, beacon_dapp_1.openToast)({
                        body: "{{wallet}}\u00A0 successfully signed payload",
                        timer: SUCCESS_TIMER,
                        state: 'finished',
                        walletInfo: data.walletInfo,
                        actions: [
                            {
                                text: "Signature: ".concat((0, shorten_string_1.shortenString)(output.signature)),
                                actionText: 'Copy to clipboard',
                                actionCallback: function () { return __awaiter(void 0, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        switch (_a.label) {
                                            case 0:
                                                navigator.clipboard.writeText(output.signature).then(function () {
                                                    logger.log('showSignSuccessAlert', 'Copying to clipboard was successful!');
                                                }, function (err) {
                                                    logger.error('showSignSuccessAlert', 'Could not copy text to clipboard: ', err);
                                                });
                                                return [4 /*yield*/, (0, beacon_dapp_1.closeToast)()];
                                            case 1:
                                                _a.sent();
                                                return [2 /*return*/];
                                        }
                                    });
                                }); }
                            }
                        ]
                    })];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
/**
 * Show a "Transaction Signed" alert
 *
 * @param data The data that is emitted by the ENCRYPT_REQUEST_SUCCESS event
 */
// TODO: ENCRYPTION
// const showEncryptSuccessAlert = async (
//   data: BeaconEventType[BeaconEvent.ENCRYPT_REQUEST_SUCCESS]
// ): Promise<void> => {
//   const output = data.output
//   await openToast({
//     body: `{{wallet}}\u00A0 successfully ${
//       data.output.cryptoOperation === EncryptionOperation.ENCRYPT ? 'encrypted' : 'decrypted'
//     } payload`,
//     timer: SUCCESS_TIMER,
//     state: 'finished',
//     walletInfo: data.walletInfo,
//     actions: [
//       {
//         text: `Payload: <strong>${shortenString(output.payload)}</strong>`,
//         actionText: 'Copy to clipboard',
//         actionCallback: async (): Promise<void> => {
//           navigator.clipboard.writeText(output.payload).then(
//             () => {
//               logger.log('showSignSuccessAlert', 'Copying to clipboard was successful!')
//             },
//             (err) => {
//               logger.error('showSignSuccessAlert', 'Could not copy text to clipboard: ', err)
//             }
//           )
//           await closeToast()
//         }
//       }
//     ]
//   })
// }
/**
 * Show a "Broadcasted" alert
 *
 * @param data The data that is emitted by the BROADCAST_REQUEST_SUCCESS event
 */
var showBroadcastSuccessAlert = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    var network, output, blockExplorer;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                network = data.network, output = data.output, blockExplorer = data.blockExplorer;
                return [4 /*yield*/, (0, beacon_dapp_1.openToast)({
                        body: "{{wallet}}\u00A0 successfully injected operation",
                        timer: SUCCESS_TIMER,
                        state: 'finished',
                        walletInfo: data.walletInfo,
                        actions: [
                            {
                                text: (0, shorten_string_1.shortenString)(output.transactionHash),
                                isBold: true,
                                actionText: "Open Blockexplorer",
                                actionLogo: 'external',
                                actionCallback: function () { return __awaiter(void 0, void 0, void 0, function () {
                                    var link;
                                    return __generator(this, function (_a) {
                                        switch (_a.label) {
                                            case 0: return [4 /*yield*/, blockExplorer.getTransactionLink(output.transactionHash, network)];
                                            case 1:
                                                link = _a.sent();
                                                window.open(link, '_blank', 'noopener');
                                                return [4 /*yield*/, (0, beacon_dapp_1.closeToast)()];
                                            case 2:
                                                _a.sent();
                                                return [2 /*return*/];
                                        }
                                    });
                                }); }
                            }
                        ]
                    })];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
var emptyHandler = function () { return function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        return [2 /*return*/];
    });
}); }; };
/**
 * The default event handlers
 */
exports.defaultEventCallbacks = (_a = {},
    _a[BeaconEvent.PERMISSION_REQUEST_SENT] = showSentToast,
    _a[BeaconEvent.PERMISSION_REQUEST_SUCCESS] = showPermissionSuccessAlert,
    _a[BeaconEvent.PERMISSION_REQUEST_ERROR] = showErrorToast,
    _a[BeaconEvent.OPERATION_REQUEST_SENT] = showSentToast,
    _a[BeaconEvent.OPERATION_REQUEST_SUCCESS] = showOperationSuccessAlert,
    _a[BeaconEvent.OPERATION_REQUEST_ERROR] = showErrorToast,
    _a[BeaconEvent.SIGN_REQUEST_SENT] = showSentToast,
    _a[BeaconEvent.SIGN_REQUEST_SUCCESS] = showSignSuccessAlert,
    _a[BeaconEvent.SIGN_REQUEST_ERROR] = showErrorToast,
    // TODO: ENCRYPTION
    // [BeaconEvent.ENCRYPT_REQUEST_SENT]: showSentToast,
    // [BeaconEvent.ENCRYPT_REQUEST_SUCCESS]: showEncryptSuccessAlert,
    // [BeaconEvent.ENCRYPT_REQUEST_ERROR]: showErrorToast,
    _a[BeaconEvent.BROADCAST_REQUEST_SENT] = showSentToast,
    _a[BeaconEvent.BROADCAST_REQUEST_SUCCESS] = showBroadcastSuccessAlert,
    _a[BeaconEvent.BROADCAST_REQUEST_ERROR] = showErrorToast,
    _a[BeaconEvent.ACKNOWLEDGE_RECEIVED] = showAcknowledgedToast,
    _a[BeaconEvent.LOCAL_RATE_LIMIT_REACHED] = showRateLimitReached,
    _a[BeaconEvent.NO_PERMISSIONS] = showNoPermissionAlert,
    _a[BeaconEvent.ACTIVE_ACCOUNT_SET] = emptyHandler(),
    _a[BeaconEvent.ACTIVE_TRANSPORT_SET] = emptyHandler(),
    _a[BeaconEvent.SHOW_PREPARE] = showPrepare,
    _a[BeaconEvent.HIDE_UI] = hideUI,
    _a[BeaconEvent.PAIR_INIT] = showPairAlert,
    _a[BeaconEvent.PAIR_SUCCESS] = showExtensionConnectedAlert,
    _a[BeaconEvent.CHANNEL_CLOSED] = showChannelClosedAlert,
    _a[BeaconEvent.INTERNAL_ERROR] = showInternalErrorAlert,
    _a[BeaconEvent.UNKNOWN] = emptyHandler(),
    _a);
/**
 * @internalapi
 *
 * Handles beacon events
 */
var BeaconEventHandler = /** @class */ (function () {
    function BeaconEventHandler(eventsToOverride, overrideAll) {
        var _a;
        if (eventsToOverride === void 0) { eventsToOverride = {}; }
        this.callbackMap = (_a = {},
            _a[BeaconEvent.PERMISSION_REQUEST_SENT] = [exports.defaultEventCallbacks.PERMISSION_REQUEST_SENT],
            _a[BeaconEvent.PERMISSION_REQUEST_SUCCESS] = [exports.defaultEventCallbacks.PERMISSION_REQUEST_SUCCESS],
            _a[BeaconEvent.PERMISSION_REQUEST_ERROR] = [exports.defaultEventCallbacks.PERMISSION_REQUEST_ERROR],
            _a[BeaconEvent.OPERATION_REQUEST_SENT] = [exports.defaultEventCallbacks.OPERATION_REQUEST_SENT],
            _a[BeaconEvent.OPERATION_REQUEST_SUCCESS] = [exports.defaultEventCallbacks.OPERATION_REQUEST_SUCCESS],
            _a[BeaconEvent.OPERATION_REQUEST_ERROR] = [exports.defaultEventCallbacks.OPERATION_REQUEST_ERROR],
            _a[BeaconEvent.SIGN_REQUEST_SENT] = [exports.defaultEventCallbacks.SIGN_REQUEST_SENT],
            _a[BeaconEvent.SIGN_REQUEST_SUCCESS] = [exports.defaultEventCallbacks.SIGN_REQUEST_SUCCESS],
            _a[BeaconEvent.SIGN_REQUEST_ERROR] = [exports.defaultEventCallbacks.SIGN_REQUEST_ERROR],
            // TODO: ENCRYPTION
            // [BeaconEvent.ENCRYPT_REQUEST_SENT]: [defaultEventCallbacks.ENCRYPT_REQUEST_SENT],
            // [BeaconEvent.ENCRYPT_REQUEST_SUCCESS]: [defaultEventCallbacks.ENCRYPT_REQUEST_SUCCESS],
            // [BeaconEvent.ENCRYPT_REQUEST_ERROR]: [defaultEventCallbacks.ENCRYPT_REQUEST_ERROR],
            _a[BeaconEvent.BROADCAST_REQUEST_SENT] = [exports.defaultEventCallbacks.BROADCAST_REQUEST_SENT],
            _a[BeaconEvent.BROADCAST_REQUEST_SUCCESS] = [exports.defaultEventCallbacks.BROADCAST_REQUEST_SUCCESS],
            _a[BeaconEvent.BROADCAST_REQUEST_ERROR] = [exports.defaultEventCallbacks.BROADCAST_REQUEST_ERROR],
            _a[BeaconEvent.ACKNOWLEDGE_RECEIVED] = [exports.defaultEventCallbacks.ACKNOWLEDGE_RECEIVED],
            _a[BeaconEvent.LOCAL_RATE_LIMIT_REACHED] = [exports.defaultEventCallbacks.LOCAL_RATE_LIMIT_REACHED],
            _a[BeaconEvent.NO_PERMISSIONS] = [exports.defaultEventCallbacks.NO_PERMISSIONS],
            _a[BeaconEvent.ACTIVE_ACCOUNT_SET] = [exports.defaultEventCallbacks.ACTIVE_ACCOUNT_SET],
            _a[BeaconEvent.ACTIVE_TRANSPORT_SET] = [exports.defaultEventCallbacks.ACTIVE_TRANSPORT_SET],
            _a[BeaconEvent.SHOW_PREPARE] = [exports.defaultEventCallbacks.SHOW_PREPARE],
            _a[BeaconEvent.HIDE_UI] = [exports.defaultEventCallbacks.HIDE_UI],
            _a[BeaconEvent.PAIR_INIT] = [exports.defaultEventCallbacks.PAIR_INIT],
            _a[BeaconEvent.PAIR_SUCCESS] = [exports.defaultEventCallbacks.PAIR_SUCCESS],
            _a[BeaconEvent.CHANNEL_CLOSED] = [exports.defaultEventCallbacks.CHANNEL_CLOSED],
            _a[BeaconEvent.INTERNAL_ERROR] = [exports.defaultEventCallbacks.INTERNAL_ERROR],
            _a[BeaconEvent.UNKNOWN] = [exports.defaultEventCallbacks.UNKNOWN],
            _a);
        if (overrideAll) {
            this.setAllHandlers();
        }
        this.overrideDefaults(eventsToOverride);
    }
    /**
     * A method to subscribe to a specific beacon event and register a callback
     *
     * @param event The event being emitted
     * @param eventCallback The callback that will be invoked
     */
    BeaconEventHandler.prototype.on = function (event, eventCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var listeners;
            return __generator(this, function (_a) {
                listeners = this.callbackMap[event] || [];
                listeners.push(eventCallback);
                this.callbackMap[event] = listeners;
                return [2 /*return*/];
            });
        });
    };
    /**
     * Emit a beacon event
     *
     * @param event The event being emitted
     * @param data The data to be emit
     */
    BeaconEventHandler.prototype.emit = function (event, data, eventCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var listeners;
            var _this = this;
            return __generator(this, function (_a) {
                listeners = this.callbackMap[event];
                if (listeners && listeners.length > 0) {
                    listeners.forEach(function (listener) { return __awaiter(_this, void 0, void 0, function () {
                        var listenerError_1;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    _a.trys.push([0, 2, , 3]);
                                    return [4 /*yield*/, listener(data, eventCallback)];
                                case 1:
                                    _a.sent();
                                    return [3 /*break*/, 3];
                                case 2:
                                    listenerError_1 = _a.sent();
                                    logger.error("error handling event ".concat(event), listenerError_1);
                                    return [3 /*break*/, 3];
                                case 3: return [2 /*return*/];
                            }
                        });
                    }); });
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Override beacon event default callbacks. This can be used to disable default alert/toast behaviour
     *
     * @param eventsToOverride An object with the events to override
     */
    BeaconEventHandler.prototype.overrideDefaults = function (eventsToOverride) {
        var _this = this;
        Object.keys(eventsToOverride).forEach(function (untypedEvent) {
            var eventType = untypedEvent;
            var event = eventsToOverride[eventType];
            if (event) {
                _this.callbackMap[eventType] = [event.handler];
            }
        });
    };
    /**
     * Set all event callbacks to a specific handler.
     */
    BeaconEventHandler.prototype.setAllHandlers = function (handler) {
        var _this = this;
        Object.keys(this.callbackMap).forEach(function (untypedEvent) {
            var eventType = untypedEvent;
            _this.callbackMap[eventType] = [];
            if (handler) {
                _this.callbackMap[eventType].push(handler);
            }
            else {
                _this.callbackMap[eventType].push(function () {
                    var data = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        data[_i] = arguments[_i];
                    }
                    logger.log.apply(logger, __spreadArray([untypedEvent], data, false));
                });
            }
        });
    };
    return BeaconEventHandler;
}());
exports.BeaconEventHandler = BeaconEventHandler;
