// @airgap/beacon-core/dist/cjs/transports/clients/CommunicationClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommunicationClient = void 0;
var beacon_utils_1 = require("@airgap/beacon-utils");
var ed25519_1 = require("@stablelib/ed25519");
var x25519_session_1 = require("@stablelib/x25519-session");
/**
 * @internalapi
 *
 *
 */
var CommunicationClient = /** @class */ (function () {
    function CommunicationClient(keyPair) {
        this.keyPair = keyPair;
    }
    /**
     * Get the public key
     */
    CommunicationClient.prototype.getPublicKey = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, (0, beacon_utils_1.toHex)((_a = this.keyPair) === null || _a === void 0 ? void 0 : _a.publicKey)];
            });
        });
    };
    /**
     * get the public key hash
     */
    CommunicationClient.prototype.getPublicKeyHash = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, beacon_utils_1.getHexHash)(this.keyPair.publicKey)];
            });
        });
    };
    /**
     * Create a cryptobox server
     *
     * @param otherPublicKey
     * @param selfKeypair
     */
    CommunicationClient.prototype.createCryptoBoxServer = function (otherPublicKey, selfKeypair) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, x25519_session_1.serverSessionKeys)({
                        publicKey: (0, ed25519_1.convertPublicKeyToX25519)(selfKeypair.publicKey),
                        secretKey: (0, ed25519_1.convertSecretKeyToX25519)(selfKeypair.secretKey)
                    }, (0, ed25519_1.convertPublicKeyToX25519)(Buffer.from(otherPublicKey, 'hex')))];
            });
        });
    };
    /**
     * Create a cryptobox client
     *
     * @param otherPublicKey
     * @param selfKeypair
     */
    CommunicationClient.prototype.createCryptoBoxClient = function (otherPublicKey, selfKeypair) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, x25519_session_1.clientSessionKeys)({
                        publicKey: (0, ed25519_1.convertPublicKeyToX25519)(selfKeypair.publicKey),
                        secretKey: (0, ed25519_1.convertSecretKeyToX25519)(selfKeypair.secretKey)
                    }, (0, ed25519_1.convertPublicKeyToX25519)(Buffer.from(otherPublicKey, 'hex')))];
            });
        });
    };
    /**
     * Encrypt a message for a specific publicKey (receiver, asymmetric)
     *
     * @param recipientPublicKey
     * @param message
     */
    CommunicationClient.prototype.encryptMessageAsymmetric = function (recipientPublicKey, message) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, beacon_utils_1.sealCryptobox)(message, Buffer.from(recipientPublicKey, 'hex'))];
            });
        });
    };
    return CommunicationClient;
}());
exports.CommunicationClient = CommunicationClient;
