// @tzkt/oazapfts/lib/module/index.js
/**
 * Utility function to handle different status codes.
 *
 * Example:
 *
 * const userId = await handle(api.register({ email, password }), {
 *   200: (user: User) => user.id,
 *   400: (err: string) => console.log(err),
 * })
 **/
export async function handle(promise, handler) {
    const { status, data } = await promise;
    const statusHandler = handler[status];
    if (statusHandler)
        return statusHandler(data);
    if (handler.default)
        return handler.default(status, data);
    throw new HttpError(status, data);
}
export const SUCCESS_CODES = [200, 201, 202, 204];
/**
 * Utility function to directly return any successful response
 * and throw a HttpError otherwise.
 *
 * Example:
 *
 * try {
 *   const userId = await ok(api.register({ email, password }));
 * }
 * catch (err) {
 *   console.log(err.status)
 * }
 */
export async function ok(promise) {
    const res = await promise;
    if (SUCCESS_CODES.some((s) => s == res.status))
        return res.data;
    throw new HttpError(res.status, res.data);
}
/**
 * Utility function to wrap an API function with `ok(...)`.
 */
export function okify(fn) {
    return (...args) => ok(fn(...args));
}
/**
 * Utility to `okify` each function of an API.
 */
export function optimistic(api) {
    const okApi = {};
    Object.entries(api).forEach(([key, value]) => {
        okApi[key] = typeof value === "function" ? okify(value) : value;
    });
    return okApi;
}
export class HttpError extends Error {
    status;
    data;
    constructor(status, data) {
        super(`Error: ${status}`);
        this.status = status;
        this.data = data;
    }
}
