// @tzkt/oazapfts/lib/module/runtime/util.js
// Encode param names and values as URIComponent
export const encodeReserved = [encodeURIComponent, encodeURIComponent];
export const allowReserved = [encodeURIComponent, encodeURI];
/**
 * Creates a tag-function to encode template strings with the given encoders.
 */
export function encode(encoders, delimiter = ",") {
    const q = (v, i) => {
        const encoder = encoders[i % encoders.length];
        if (typeof v === "undefined") {
            return "";
        }
        if (typeof v === "object") {
            if (Array.isArray(v)) {
                return v.map(encoder).join(delimiter);
            }
            const flat = Object.entries(v).reduce((flat, entry) => [...flat, ...entry], []);
            return flat.map(encoder).join(delimiter);
        }
        return encoder(String(v));
    };
    return (strings, ...values) => {
        return strings.reduce((prev, s, i) => {
            return `${prev}${s}${q(values[i], i)}`;
        }, "");
    };
}
/**
 * Separate array values by the given delimiter.
 */
export function delimited(delimiter = ",") {
    return (params, encoders = encodeReserved) => Object.entries(params)
        .filter(([, value]) => value !== undefined)
        .map(([name, value]) => encode(encoders, delimiter) `${name}=${value}`)
        .join("&");
}
/**
 * Deeply remove all properties with undefined values.
 */
export function stripUndefined(obj) {
    return obj && JSON.parse(JSON.stringify(obj));
}
export function joinUrl(...parts) {
    return parts
        .filter(Boolean)
        .map((s, i) => (i === 0 ? s : s.replace(/^\/+/, "")))
        .map((s, i, a) => (i === a.length - 1 ? s : s.replace(/\/+$/, "")))
        .join("/");
}
