// @taquito/taquito/dist/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublicKeyNotFoundError = exports.InvalidPrepareParamsError = exports.RPCResponseError = exports.InvalidFilterExpressionError = exports.ConfirmationUndefinedError = exports.InvalidConfirmationCountError = void 0;
/**
 *  @category Error
 *  @description Error that indicates invalid confirmation count has been passed or configured
 */
class InvalidConfirmationCountError extends Error {
    constructor(message) {
        super(message);
        this.message = message;
        this.name = 'InvalidConfirmationCountError';
    }
}
exports.InvalidConfirmationCountError = InvalidConfirmationCountError;
/**
 *  @category Error
 *  @description Error that indicates undefined confirmation has not been specified or configured
 */
class ConfirmationUndefinedError extends Error {
    constructor(message) {
        super(message);
        this.message = message;
        this.name = 'ConfirmationUndefinedError';
    }
}
exports.ConfirmationUndefinedError = ConfirmationUndefinedError;
/**
 *  @category Error
 *  @description Error that indicates an invalid filter expression being passed or used
 */
class InvalidFilterExpressionError extends Error {
    constructor(message) {
        super(message);
        this.message = message;
        this.name = 'InvalidFilterExpressionError';
    }
}
exports.InvalidFilterExpressionError = InvalidFilterExpressionError;
/**
 *  @category Error
 *  @description Error that indicates an error being returned from the RPC response
 */
class RPCResponseError extends Error {
    constructor(message) {
        super(message);
        this.message = message;
        this.name = 'RPCResponseError';
    }
}
exports.RPCResponseError = RPCResponseError;
/**
 *  @category Error
 *  @description Error that indicates invalid Preparation parameters being passed
 */
class InvalidPrepareParamsError extends Error {
    constructor(opKind) {
        super(`No '${opKind}' operation parameters have been passed`);
        this.opKind = opKind;
        this.name = 'InvalidOperationParamsError';
    }
}
exports.InvalidPrepareParamsError = InvalidPrepareParamsError;
/**
 *  @category Error
 *  @description Error that indicates invalid Preparation parameters being passed
 */
class PublicKeyNotFoundError extends Error {
    constructor() {
        super(`Unable to retrieve public key from signer. If you are using a wallet, make sure your account is revealed`);
        this.name = 'PublicKeyNotFoundError';
    }
}
exports.PublicKeyNotFoundError = PublicKeyNotFoundError;
