// @taquito/http-utils/dist/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpResponseError = exports.HttpRequestFailed = void 0;
const core_1 = require("@taquito/core");
/**
 *  @category Error
 *  @description Error that indicates a general failure in making the HTTP request
 */
class HttpRequestFailed extends core_1.NetworkError {
    constructor(method, url, cause) {
        super();
        this.method = method;
        this.url = url;
        this.cause = cause;
        this.name = 'HttpRequestFailed';
        this.message = `${method} ${url} ${String(cause)}`;
    }
}
exports.HttpRequestFailed = HttpRequestFailed;
/**
 *  @category Error
 *  @description Error thrown when the endpoint returns an HTTP error to the client
 */
class HttpResponseError extends core_1.NetworkError {
    constructor(message, status, statusText, body, url) {
        super();
        this.message = message;
        this.status = status;
        this.statusText = statusText;
        this.body = body;
        this.url = url;
        this.name = 'HttpResponse';
    }
}
exports.HttpResponseError = HttpResponseError;
