// @taquito/taquito/dist/lib/subscribe/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidFilterExpressionError = exports.UnsupportedEventError = void 0;
const core_1 = require("@taquito/core");
/**
 *  @category Error
 *  @description Error that indicates an unsupported event being passed or used
 */
class UnsupportedEventError extends core_1.ParameterValidationError {
    constructor(type) {
        super();
        this.type = type;
        this.name = 'UnsupportedEventError';
        this.message = `Unsupported event type "${type}" expecting one of the "data", "error", or "close".`;
    }
}
exports.UnsupportedEventError = UnsupportedEventError;
/**
 *  @category Error
 *  @description Error indicates an invalid filter expression being passed or used
 */
class InvalidFilterExpressionError extends core_1.ParameterValidationError {
    constructor(invalidExpression) {
        super();
        this.invalidExpression = invalidExpression;
        this.name = 'InvalidFilterExpressionError';
        this.message = `Invalid filter expression expecting the object to contain either and/or property`;
    }
}
exports.InvalidFilterExpressionError = InvalidFilterExpressionError;
