// @stablelib/x25519-session/lib/keyagreement.js
"use strict";
// Copyright (C) 2020 Dmitry Chestnykh
// MIT License. See LICENSE file for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.X25519Session = exports.SECRET_SEED_LENGTH = exports.SAVED_STATE_LENGTH = exports.ACCEPT_MESSAGE_LENGTH = exports.OFFER_MESSAGE_LENGTH = void 0;
const random_1 = require("@stablelib/random");
const wipe_1 = require("@stablelib/wipe");
const x25519_1 = require("@stablelib/x25519");
const x25519_session_1 = require("./x25519-session");
/** Constants for key agreement */
exports.OFFER_MESSAGE_LENGTH = x25519_1.PUBLIC_KEY_LENGTH;
exports.ACCEPT_MESSAGE_LENGTH = x25519_1.PUBLIC_KEY_LENGTH;
exports.SAVED_STATE_LENGTH = x25519_1.SECRET_KEY_LENGTH;
exports.SECRET_SEED_LENGTH = x25519_1.SECRET_KEY_LENGTH;
/**
 * X25519 key agreement using ephemeral key pairs.
 *
 * Note that unless this key agreement is combined with an authentication
 * method, such as public key signatures, it's vulnerable to man-in-the-middle
 * attacks.
 */
class X25519Session {
    constructor(secretSeed, prng) {
        this.offerMessageLength = exports.OFFER_MESSAGE_LENGTH;
        this.acceptMessageLength = exports.ACCEPT_MESSAGE_LENGTH;
        this.sharedKeyLength = x25519_1.SHARED_KEY_LENGTH;
        this.savedStateLength = exports.SAVED_STATE_LENGTH;
        this._seed = secretSeed || (0, random_1.randomBytes)(x25519_1.SECRET_KEY_LENGTH, prng);
    }
    saveState() {
        return new Uint8Array(this._seed);
    }
    restoreState(savedState) {
        this._seed = new Uint8Array(savedState);
        return this;
    }
    clean() {
        if (this._seed) {
            (0, wipe_1.wipe)(this._seed);
        }
        if (this._keyPair) {
            (0, wipe_1.wipe)(this._keyPair.secretKey);
            (0, wipe_1.wipe)(this._keyPair.publicKey);
        }
        if (this._sharedKey) {
            (0, wipe_1.wipe)(this._sharedKey);
        }
        if (this._sessionKeys) {
            (0, wipe_1.wipe)(this._sessionKeys.receive);
            (0, wipe_1.wipe)(this._sessionKeys.send);
        }
    }
    offer() {
        this._keyPair = (0, x25519_1.generateKeyPairFromSeed)(this._seed);
        return new Uint8Array(this._keyPair.publicKey);
    }
    accept(offerMsg) {
        if (this._keyPair) {
            throw new Error("X25519Session: accept shouldn't be called by offering party");
        }
        if (offerMsg.length !== this.offerMessageLength) {
            throw new Error("X25519Session: incorrect offer message length");
        }
        if (this._sharedKey) {
            throw new Error("X25519Session: accept was already called");
        }
        const keyPair = (0, x25519_1.generateKeyPairFromSeed)(this._seed);
        this._sharedKey = (0, x25519_1.sharedKey)(keyPair.secretKey, offerMsg);
        this._sessionKeys = (0, x25519_session_1.clientSessionKeysFromSharedKey)(this._sharedKey, keyPair.publicKey, offerMsg);
        (0, wipe_1.wipe)(keyPair.secretKey);
        return keyPair.publicKey;
    }
    finish(acceptMsg) {
        if (acceptMsg.length !== this.acceptMessageLength) {
            throw new Error("X25519Session: incorrect accept message length");
        }
        if (!this._keyPair) {
            throw new Error("X25519Session: no offer state");
        }
        if (this._sharedKey) {
            throw new Error("X25519Session: finish was already called");
        }
        this._sharedKey = (0, x25519_1.sharedKey)(this._keyPair.secretKey, acceptMsg);
        this._sessionKeys = (0, x25519_session_1.serverSessionKeysFromSharedKey)(this._sharedKey, this._keyPair.publicKey, acceptMsg);
        return this;
    }
    getSharedKey() {
        if (!this._sharedKey) {
            throw new Error("X25519Session: no shared key established");
        }
        return new Uint8Array(this._sharedKey);
    }
    getSessionKeys() {
        if (!this._sessionKeys) {
            throw new Error("X25519Session: no shared key established");
        }
        return {
            receive: new Uint8Array(this._sessionKeys.receive),
            send: new Uint8Array(this._sessionKeys.send),
        };
    }
}
exports.X25519Session = X25519Session;
