// @taqueria/node-sdk/chunk-7FZZQM3P.mjs
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw new Error('Dynamic require of "' + x + '" is not supported');
});

// types.ts
import * as Protocol from "@taqueria/protocol";
import * as Environment from "@taqueria/protocol/Environment";
import * as LoadedConfig from "@taqueria/protocol/LoadedConfig";
import * as MetadataConfig from "@taqueria/protocol/MetadataConfig";
import * as NetworkConfig from "@taqueria/protocol/NetworkConfig";
import * as Operation from "@taqueria/protocol/Operation";
import * as Option from "@taqueria/protocol/Option";
import * as PersistentState from "@taqueria/protocol/PersistentState";
import * as PluginSchema from "@taqueria/protocol/PluginSchema";
import * as PositionalArg from "@taqueria/protocol/PositionalArg";
import * as SandboxAccountConfig from "@taqueria/protocol/SandboxAccountConfig";
import * as SandboxConfig from "@taqueria/protocol/SandboxConfig";
import * as SanitizedAbsPath from "@taqueria/protocol/SanitizedAbsPath";
import * as SanitizedPath from "@taqueria/protocol/SanitizedPath";
import * as TaqError from "@taqueria/protocol/TaqError";
import * as Task from "@taqueria/protocol/Task";
import * as Template from "@taqueria/protocol/Template";

export {
  __require,
  Protocol,
  Environment,
  LoadedConfig,
  MetadataConfig,
  NetworkConfig,
  Operation,
  Option,
  PersistentState,
  PluginSchema,
  PositionalArg,
  SandboxAccountConfig,
  SandboxConfig,
  SanitizedAbsPath,
  SanitizedPath,
  TaqError,
  Task,
  Template
};
