// @taqueria/node-sdk/types.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// types.ts
var types_exports = {};
__export(types_exports, {
  Environment: () => Environment,
  LoadedConfig: () => LoadedConfig,
  MetadataConfig: () => MetadataConfig,
  NetworkConfig: () => NetworkConfig,
  Operation: () => Operation,
  Option: () => Option,
  PersistentState: () => PersistentState,
  PluginSchema: () => PluginSchema,
  PositionalArg: () => PositionalArg,
  Protocol: () => Protocol,
  SandboxAccountConfig: () => SandboxAccountConfig,
  SandboxConfig: () => SandboxConfig,
  SanitizedAbsPath: () => SanitizedAbsPath,
  SanitizedPath: () => SanitizedPath,
  TaqError: () => TaqError,
  Task: () => Task,
  Template: () => Template
});
module.exports = __toCommonJS(types_exports);
var Protocol = __toESM(require("@taqueria/protocol"));
var Environment = __toESM(require("@taqueria/protocol/Environment"));
var LoadedConfig = __toESM(require("@taqueria/protocol/LoadedConfig"));
var MetadataConfig = __toESM(require("@taqueria/protocol/MetadataConfig"));
var NetworkConfig = __toESM(require("@taqueria/protocol/NetworkConfig"));
var Operation = __toESM(require("@taqueria/protocol/Operation"));
var Option = __toESM(require("@taqueria/protocol/Option"));
var PersistentState = __toESM(require("@taqueria/protocol/PersistentState"));
var PluginSchema = __toESM(require("@taqueria/protocol/PluginSchema"));
var PositionalArg = __toESM(require("@taqueria/protocol/PositionalArg"));
var SandboxAccountConfig = __toESM(require("@taqueria/protocol/SandboxAccountConfig"));
var SandboxConfig = __toESM(require("@taqueria/protocol/SandboxConfig"));
var SanitizedAbsPath = __toESM(require("@taqueria/protocol/SanitizedAbsPath"));
var SanitizedPath = __toESM(require("@taqueria/protocol/SanitizedPath"));
var TaqError = __toESM(require("@taqueria/protocol/TaqError"));
var Task = __toESM(require("@taqueria/protocol/Task"));
var Template = __toESM(require("@taqueria/protocol/Template"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Environment,
  LoadedConfig,
  MetadataConfig,
  NetworkConfig,
  Operation,
  Option,
  PersistentState,
  PluginSchema,
  PositionalArg,
  Protocol,
  SandboxAccountConfig,
  SandboxConfig,
  SanitizedAbsPath,
  SanitizedPath,
  TaqError,
  Task,
  Template
});
