// @airgap/beacon-dapp/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TezblockBlockExplorer = exports.TzktBlockExplorer = exports.BlockExplorer = exports.defaultEventCallbacks = exports.BeaconEventHandler = exports.BeaconEvent = exports.getDAppClientInstance = exports.DAppClient = void 0;
__exportStar(require("@airgap/beacon-core"), exports);
__exportStar(require("@airgap/beacon-transport-matrix"), exports);
__exportStar(require("@airgap/beacon-transport-postmessage"), exports);
__exportStar(require("@airgap/beacon-types"), exports);
__exportStar(require("@airgap/beacon-utils"), exports);
__exportStar(require("@airgap/beacon-ui"), exports);
var DAppClient_1 = require("./dapp-client/DAppClient");
Object.defineProperty(exports, "DAppClient", { enumerable: true, get: function () { return DAppClient_1.DAppClient; } });
var events_1 = require("./events");
Object.defineProperty(exports, "BeaconEvent", { enumerable: true, get: function () { return events_1.BeaconEvent; } });
Object.defineProperty(exports, "BeaconEventHandler", { enumerable: true, get: function () { return events_1.BeaconEventHandler; } });
Object.defineProperty(exports, "defaultEventCallbacks", { enumerable: true, get: function () { return events_1.defaultEventCallbacks; } });
var block_explorer_1 = require("./utils/block-explorer");
Object.defineProperty(exports, "BlockExplorer", { enumerable: true, get: function () { return block_explorer_1.BlockExplorer; } });
var tzkt_blockexplorer_1 = require("./utils/tzkt-blockexplorer");
Object.defineProperty(exports, "TzktBlockExplorer", { enumerable: true, get: function () { return tzkt_blockexplorer_1.TzktBlockExplorer; } });
Object.defineProperty(exports, "TezblockBlockExplorer", { enumerable: true, get: function () { return tzkt_blockexplorer_1.TzktBlockExplorer; } });
var get_instance_1 = require("./utils/get-instance");
Object.defineProperty(exports, "getDAppClientInstance", { enumerable: true, get: function () { return get_instance_1.getDAppClientInstance; } });
