// @taquito/signer/dist/lib/derivation-tools/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Path = exports.Hard = exports.Ed25519 = exports.ECDSA = void 0;
const errors_1 = require("../errors");
exports.ECDSA = require("./ecdsa");
exports.Ed25519 = require("./ed25519");
exports.Hard = 0x80000000;
class Path extends Array {
    static from(iterable) {
        return super.from(iterable).map((x) => x >>> 0);
    }
    /**
     *
     * @param s derivation path eg: 44'/1729'/0'/0'
     * @returns applied hardened values
     */
    static fromString(s) {
        if (s.length === 0) {
            return new Path();
        }
        let parts = s.split('/');
        const out = [];
        if (parts[0] === 'm') {
            parts = parts.slice(1);
        }
        for (let p of parts) {
            if (p.length === 0) {
                throw new errors_1.InvalidDerivationPathError(`invalid BIP32 path: ${s}`);
            }
            let h = 0;
            const last = p[p.length - 1];
            if (last === "'" || last === 'h' || last === 'H') {
                h = exports.Hard;
                p = p.slice(0, p.length - 1);
            }
            const index = (parseInt(p, 10) | h) >>> 0;
            out.push(index);
        }
        return Path.from(out);
    }
}
exports.Path = Path;
