// @airgap/beacon-ui/dist/index/components/pairing.jsx
import { desktopList, extensionList, iOSList, webList } from '../ui/alert/wallet-lists';
// export interface PairingProps {}
// const Pairing: Component<PairingProps> = (props: PairingProps) => {
//   return (
//     <div>
//       <h1>Pairing Component</h1>
//     </div>
//   )
// }
// export default Pairing
/**
 * Initialize with tezos wallets for backwards compatibility
 */
let localDesktopList = desktopList;
let localExtensionList = extensionList;
let localWebList = webList;
let localiOSList = iOSList;
export const getDesktopList = () => {
    return localDesktopList;
};
export const setDesktopList = (desktopList) => {
    localDesktopList = desktopList;
};
export const getExtensionList = () => {
    return localExtensionList;
};
export const setExtensionList = (extensionList) => {
    localExtensionList = extensionList;
};
export const getWebList = () => {
    return localWebList;
};
export const setWebList = (webList) => {
    localWebList = webList;
};
export const getiOSList = () => {
    return localiOSList;
};
export const setiOSList = (iosList) => {
    localiOSList = iosList;
};
export var Platform;
(function (Platform) {
    Platform[Platform["DESKTOP"] = 0] = "DESKTOP";
    Platform[Platform["IOS"] = 1] = "IOS";
    Platform[Platform["ANDROID"] = 2] = "ANDROID";
})(Platform || (Platform = {}));
export var WalletType;
(function (WalletType) {
    WalletType["IOS"] = "ios";
    WalletType["ANDROID"] = "android";
    WalletType["EXTENSION"] = "extension";
    WalletType["DESKTOP"] = "desktop";
    WalletType["WEB"] = "web";
})(WalletType || (WalletType = {}));
/**
 * @internalapi
 *
 */
export class Pairing {
}
