// @airgap/beacon-transport-walletconnect/dist/cjs/communication-client/WalletConnectCommunicationClient.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletConnectCommunicationClient = exports.PermissionScopeEvents = exports.PermissionScopeMethods = void 0;
var beacon_core_1 = require("@airgap/beacon-core");
var sign_client_1 = require("@walletconnect/sign-client");
var utils_1 = require("@walletconnect/utils");
var error_1 = require("../error");
var beacon_types_1 = require("@airgap/beacon-types");
var beacon_utils_1 = require("@airgap/beacon-utils");
var TEZOS_PLACEHOLDER = 'tezos';
var PermissionScopeMethods;
(function (PermissionScopeMethods) {
    PermissionScopeMethods["GET_ACCOUNTS"] = "tezos_getAccounts";
    PermissionScopeMethods["OPERATION_REQUEST"] = "tezos_send";
    PermissionScopeMethods["SIGN"] = "tezos_sign";
})(PermissionScopeMethods = exports.PermissionScopeMethods || (exports.PermissionScopeMethods = {}));
var PermissionScopeEvents;
(function (PermissionScopeEvents) {
    PermissionScopeEvents["CHAIN_CHANGED"] = "chainChanged";
    PermissionScopeEvents["ACCOUNTS_CHANGED"] = "accountsChanged";
    PermissionScopeEvents["REQUEST_ACKNOWLEDGED"] = "requestAcknowledged";
})(PermissionScopeEvents = exports.PermissionScopeEvents || (exports.PermissionScopeEvents = {}));
var WalletConnectCommunicationClient = /** @class */ (function (_super) {
    __extends(WalletConnectCommunicationClient, _super);
    function WalletConnectCommunicationClient(wcOptions) {
        var _this = _super.call(this) || this;
        _this.wcOptions = wcOptions;
        _this.activeListeners = new Map();
        _this.channelOpeningListeners = new Map();
        _this.getSignClient();
        return _this;
    }
    WalletConnectCommunicationClient.getInstance = function (wcOptions) {
        if (!WalletConnectCommunicationClient.instance) {
            WalletConnectCommunicationClient.instance = new WalletConnectCommunicationClient(wcOptions);
        }
        return WalletConnectCommunicationClient.instance;
    };
    WalletConnectCommunicationClient.prototype.listenForEncryptedMessage = function (senderPublicKey, messageCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var callbackFunction;
            var _this = this;
            return __generator(this, function (_a) {
                if (this.activeListeners.has(senderPublicKey)) {
                    return [2 /*return*/];
                }
                callbackFunction = function (message) { return __awaiter(_this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        messageCallback(message);
                        return [2 /*return*/];
                    });
                }); };
                this.activeListeners.set(senderPublicKey, callbackFunction);
                return [2 /*return*/];
            });
        });
    };
    WalletConnectCommunicationClient.prototype.listenForChannelOpening = function (messageCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var callbackFunction;
            var _this = this;
            return __generator(this, function (_a) {
                callbackFunction = function (pairingResponse) { return __awaiter(_this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        messageCallback(pairingResponse);
                        return [2 /*return*/];
                    });
                }); };
                this.channelOpeningListeners.set('channelOpening', callbackFunction);
                return [2 /*return*/];
            });
        });
    };
    WalletConnectCommunicationClient.prototype.unsubscribeFromEncryptedMessages = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    WalletConnectCommunicationClient.prototype.unsubscribeFromEncryptedMessage = function (_senderPublicKey) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    WalletConnectCommunicationClient.prototype.sendMessage = function (_message, _peer) {
        return __awaiter(this, void 0, void 0, function () {
            var serializer, message;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        serializer = new beacon_core_1.Serializer();
                        return [4 /*yield*/, serializer.deserialize(_message)];
                    case 1:
                        message = (_a.sent());
                        if (!message) {
                            return [2 /*return*/];
                        }
                        this.currentMessageId = message.id;
                        switch (message.type) {
                            case beacon_types_1.BeaconMessageType.PermissionRequest:
                                this.requestPermissions(message);
                                break;
                            case beacon_types_1.BeaconMessageType.OperationRequest:
                                this.sendOperations(message);
                                break;
                            case beacon_types_1.BeaconMessageType.SignPayloadRequest:
                                this.signPayload(message);
                                break;
                            default:
                                return [2 /*return*/];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.fetchAccounts = function (topic, chainId) {
        return __awaiter(this, void 0, void 0, function () {
            var signClient;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getSignClient()];
                    case 1:
                        signClient = _a.sent();
                        return [2 /*return*/, signClient.request({
                                topic: topic,
                                chainId: chainId,
                                request: {
                                    method: PermissionScopeMethods.GET_ACCOUNTS,
                                    params: {}
                                }
                            })];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.requestPermissions = function (message) {
        var _a, _b, _c, _d, _e;
        return __awaiter(this, void 0, void 0, function () {
            var error_2, session, publicKey, accounts, addressOrPbk, result, permissionResponse;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        console.log('#### Requesting permissions');
                        if (!this.getPermittedMethods().includes(PermissionScopeMethods.GET_ACCOUNTS)) {
                            throw new error_1.MissingRequiredScope(PermissionScopeMethods.GET_ACCOUNTS);
                        }
                        if (!this.activeAccount) return [3 /*break*/, 4];
                        _f.label = 1;
                    case 1:
                        _f.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.openSession()];
                    case 2:
                        _f.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        error_2 = _f.sent();
                        console.error(error_2.message);
                        return [2 /*return*/];
                    case 4:
                        this.setDefaultAccountAndNetwork();
                        session = this.getSession();
                        if (!(((_a = session.sessionProperties) === null || _a === void 0 ? void 0 : _a.pubkey) &&
                            ((_b = session.sessionProperties) === null || _b === void 0 ? void 0 : _b.algo) &&
                            ((_c = session.sessionProperties) === null || _c === void 0 ? void 0 : _c.address))) return [3 /*break*/, 5];
                        publicKey = (_d = session.sessionProperties) === null || _d === void 0 ? void 0 : _d.pubkey;
                        console.log('[requestPermissions]: Have pubkey in sessionProperties, skipping "get_accounts" call', session.sessionProperties);
                        return [3 /*break*/, 8];
                    case 5:
                        accounts = this.getTezosNamespace(session.namespaces).accounts;
                        addressOrPbk = accounts[0].split(':', 3)[2];
                        if (!addressOrPbk.startsWith('edpk')) return [3 /*break*/, 6];
                        publicKey = addressOrPbk;
                        return [3 /*break*/, 8];
                    case 6:
                        if (message.network.type !== this.wcOptions.network) {
                            throw new Error('Network in permission request is not the same as preferred network!');
                        }
                        return [4 /*yield*/, this.fetchAccounts(session.topic, "".concat(TEZOS_PLACEHOLDER, ":").concat(message.network.type))];
                    case 7:
                        result = _f.sent();
                        if (!result || result.length < 1) {
                            throw new Error('No account shared by wallet');
                        }
                        if (result.some(function (account) { return !account.pubkey; })) {
                            throw new Error('Public Key in `tezos_getAccounts` is empty!');
                        }
                        publicKey = (_e = result[0]) === null || _e === void 0 ? void 0 : _e.pubkey;
                        _f.label = 8;
                    case 8:
                        if (!publicKey) {
                            throw new Error('Public Key in `tezos_getAccounts` is empty!');
                        }
                        permissionResponse = {
                            type: beacon_types_1.BeaconMessageType.PermissionResponse,
                            appMetadata: {
                                senderId: session.pairingTopic,
                                name: session.peer.metadata.name,
                                icon: session.peer.metadata.icons[0]
                            },
                            publicKey: publicKey,
                            network: message.network,
                            scopes: [beacon_types_1.PermissionScope.SIGN, beacon_types_1.PermissionScope.OPERATION_REQUEST],
                            id: this.currentMessageId
                        };
                        this.notifyListeners(session.pairingTopic, permissionResponse);
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @description Once the session is establish, send payload to be approved and signed by the wallet.
     * @error MissingRequiredScope is thrown if permission to sign payload was not granted
     */
    WalletConnectCommunicationClient.prototype.signPayload = function (signPayloadRequest) {
        return __awaiter(this, void 0, void 0, function () {
            var signClient, session, network, account;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getSignClient()];
                    case 1:
                        signClient = _a.sent();
                        session = this.getSession();
                        if (!this.getPermittedMethods().includes(PermissionScopeMethods.SIGN)) {
                            throw new error_1.MissingRequiredScope(PermissionScopeMethods.SIGN);
                        }
                        network = this.getActiveNetwork();
                        return [4 /*yield*/, this.getPKH()];
                    case 2:
                        account = _a.sent();
                        this.validateNetworkAndAccount(network, account);
                        // TODO: Type
                        signClient
                            .request({
                            topic: session.topic,
                            chainId: "".concat(TEZOS_PLACEHOLDER, ":").concat(network),
                            request: {
                                method: PermissionScopeMethods.SIGN,
                                params: {
                                    account: account,
                                    payload: signPayloadRequest.payload
                                }
                            }
                        })
                            .then(function (response) {
                            var signPayloadResponse = {
                                type: beacon_types_1.BeaconMessageType.SignPayloadResponse,
                                signingType: signPayloadRequest.signingType,
                                signature: response === null || response === void 0 ? void 0 : response.signature,
                                id: _this.currentMessageId
                            };
                            _this.notifyListeners(session.pairingTopic, signPayloadResponse);
                        })
                            .catch(function () { return __awaiter(_this, void 0, void 0, function () {
                            var errorResponse;
                            return __generator(this, function (_a) {
                                errorResponse = {
                                    type: beacon_types_1.BeaconMessageType.Error,
                                    id: this.currentMessageId,
                                    errorType: beacon_types_1.BeaconErrorType.ABORTED_ERROR
                                };
                                this.notifyListeners(session.pairingTopic, errorResponse);
                                return [2 /*return*/];
                            });
                        }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @description Once the session is established, send Tezos operations to be approved, signed and inject by the wallet.
     * @error MissingRequiredScope is thrown if permission to send operation was not granted
     */
    WalletConnectCommunicationClient.prototype.sendOperations = function (operationRequest) {
        return __awaiter(this, void 0, void 0, function () {
            var signClient, session, network, account;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getSignClient()];
                    case 1:
                        signClient = _a.sent();
                        session = this.getSession();
                        if (!this.getPermittedMethods().includes(PermissionScopeMethods.OPERATION_REQUEST)) {
                            throw new error_1.MissingRequiredScope(PermissionScopeMethods.OPERATION_REQUEST);
                        }
                        network = this.getActiveNetwork();
                        return [4 /*yield*/, this.getPKH()];
                    case 2:
                        account = _a.sent();
                        this.validateNetworkAndAccount(network, account);
                        signClient
                            .request({
                            topic: session.topic,
                            chainId: "".concat(TEZOS_PLACEHOLDER, ":").concat(network),
                            request: {
                                method: PermissionScopeMethods.OPERATION_REQUEST,
                                params: {
                                    account: account,
                                    operations: operationRequest.operationDetails
                                }
                            }
                        })
                            .then(function (response) {
                            var _a, _b, _c;
                            var sendOperationResponse = {
                                type: beacon_types_1.BeaconMessageType.OperationResponse,
                                transactionHash: (_c = (_b = (_a = response.operationHash) !== null && _a !== void 0 ? _a : response.transactionHash) !== null && _b !== void 0 ? _b : response.hash) !== null && _c !== void 0 ? _c : '',
                                id: _this.currentMessageId
                            };
                            _this.notifyListeners(session.pairingTopic, sendOperationResponse);
                        })
                            .catch(function () { return __awaiter(_this, void 0, void 0, function () {
                            var errorResponse;
                            return __generator(this, function (_a) {
                                errorResponse = {
                                    type: beacon_types_1.BeaconMessageType.Error,
                                    id: this.currentMessageId,
                                    errorType: beacon_types_1.BeaconErrorType.ABORTED_ERROR
                                };
                                this.notifyListeners(session.pairingTopic, errorResponse);
                                return [2 /*return*/];
                            });
                        }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.init = function (forceNewConnection) {
        if (forceNewConnection === void 0) { forceNewConnection = false; }
        return __awaiter(this, void 0, void 0, function () {
            var signClient, sessions, _a, uri, topic;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getSignClient()];
                    case 1:
                        signClient = _b.sent();
                        if (forceNewConnection) {
                            this.closePairings();
                        }
                        sessions = signClient.session.getAll();
                        if (sessions && sessions.length > 0) {
                            this.session = sessions[0];
                            this.setDefaultAccountAndNetwork();
                            return [2 /*return*/, undefined];
                        }
                        return [4 /*yield*/, signClient.core.pairing.create()];
                    case 2:
                        _a = _b.sent(), uri = _a.uri, topic = _a.topic;
                        signClient.core.pairing.ping({ topic: topic }).then(function () { return __awaiter(_this, void 0, void 0, function () {
                            var session, pairingResponse_1, error_3, fun;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, signClient.core.pairing.activate({ topic: topic })
                                        // pairings don't have peer details
                                        // therefore we must immediately open a session
                                        // to get data required in the pairing response
                                    ];
                                    case 1:
                                        _a.sent();
                                        _a.label = 2;
                                    case 2:
                                        _a.trys.push([2, 4, , 5]);
                                        return [4 /*yield*/, this.openSession(topic)];
                                    case 3:
                                        session = _a.sent();
                                        pairingResponse_1 = {
                                            id: topic,
                                            type: 'walletconnect-pairing-response',
                                            name: session.peer.metadata.name,
                                            publicKey: session.peer.publicKey,
                                            senderId: topic,
                                            extensionId: session.peer.metadata.name,
                                            version: '3'
                                        };
                                        this.channelOpeningListeners.forEach(function (listener) {
                                            listener(pairingResponse_1);
                                        });
                                        return [3 /*break*/, 5];
                                    case 4:
                                        error_3 = _a.sent();
                                        console.error(error_3.message);
                                        fun = this.eventHandlers.get("CLOSE_ALERT" /* ClientEvents.CLOSE_ALERT */);
                                        fun && fun();
                                        return [2 /*return*/];
                                    case 5: return [2 /*return*/];
                                }
                            });
                        }); });
                        return [2 /*return*/, { uri: uri, topic: topic }];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.close = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.closePairings()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.subscribeToSessionEvents = function (signClient) {
        var _this = this;
        signClient.on('session_event', function (event) {
            if (event.params.event.name === PermissionScopeEvents.REQUEST_ACKNOWLEDGED &&
                _this.currentMessageId) {
                _this.acknowledgeRequest(_this.currentMessageId);
            }
        });
        signClient.on('session_update', function (event) {
            _this.updateActiveAccount(event.params.namespaces);
        });
        signClient.on('session_delete', function (event) {
            _this.disconnect(signClient, { type: 'session', topic: event.topic });
        });
        signClient.on('session_expire', function (event) {
            _this.disconnect(signClient, { type: 'session', topic: event.topic });
        });
        signClient.core.pairing.events.on('pairing_delete', function (event) {
            _this.disconnect(signClient, { type: 'pairing', topic: event.topic });
        });
    };
    WalletConnectCommunicationClient.prototype.acknowledgeRequest = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var session, acknowledgeResponse;
            return __generator(this, function (_a) {
                session = this.getSession();
                acknowledgeResponse = {
                    type: beacon_types_1.BeaconMessageType.Acknowledge,
                    id: id
                };
                this.notifyListeners(session.pairingTopic, acknowledgeResponse);
                return [2 /*return*/];
            });
        });
    };
    WalletConnectCommunicationClient.prototype.updateActiveAccount = function (namespaces) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var accounts, _b, _namespace, chainId, addressOrPbk_1, session, publicKey, _c, result, _d, _e, _f;
            var _g;
            return __generator(this, function (_h) {
                switch (_h.label) {
                    case 0:
                        _h.trys.push([0, 7, , 8]);
                        accounts = this.getTezosNamespace(namespaces).accounts;
                        if (!(accounts.length === 1)) return [3 /*break*/, 6];
                        _b = accounts[0].split(':', 3), _namespace = _b[0], chainId = _b[1], addressOrPbk_1 = _b[2];
                        session = this.getSession();
                        publicKey = void 0;
                        this.activeNetwork = chainId;
                        if (!addressOrPbk_1.startsWith('edpk')) return [3 /*break*/, 2];
                        publicKey = addressOrPbk_1;
                        _c = this;
                        return [4 /*yield*/, (0, beacon_utils_1.getAddressFromPublicKey)(publicKey)];
                    case 1:
                        _c.activeAccount = _h.sent();
                        return [3 /*break*/, 4];
                    case 2:
                        this.activeAccount = addressOrPbk_1;
                        return [4 /*yield*/, this.fetchAccounts(session.topic, "".concat(TEZOS_PLACEHOLDER, ":").concat(chainId))];
                    case 3:
                        result = _h.sent();
                        publicKey = (_a = result === null || result === void 0 ? void 0 : result.find(function (_a) {
                            var _address = _a.address;
                            return addressOrPbk_1 === _address;
                        })) === null || _a === void 0 ? void 0 : _a.pubkey;
                        _h.label = 4;
                    case 4:
                        if (!publicKey) {
                            throw new Error('Public key for the new account not provided');
                        }
                        _d = this.notifyListeners;
                        _e = [session.pairingTopic];
                        _g = {};
                        return [4 /*yield*/, (0, beacon_utils_1.generateGUID)()];
                    case 5:
                        _d.apply(this, _e.concat([(_g.id = _h.sent(),
                                _g.type = beacon_types_1.BeaconMessageType.ChangeAccountRequest,
                                _g.publicKey = publicKey,
                                _g.network = { type: chainId },
                                _g.scopes = [beacon_types_1.PermissionScope.SIGN, beacon_types_1.PermissionScope.OPERATION_REQUEST],
                                _g)]));
                        _h.label = 6;
                    case 6: return [3 /*break*/, 8];
                    case 7:
                        _f = _h.sent();
                        return [3 /*break*/, 8];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.disconnect = function (signClient, trigger) {
        return __awaiter(this, void 0, void 0, function () {
            var session, fun, _a, _b;
            var _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (!(trigger.type === 'session')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.onSessionClosed(signClient, trigger.topic)];
                    case 1:
                        session = _d.sent();
                        _d.label = 2;
                    case 2:
                        if (!(trigger.type === 'pairing')) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.onPairingClosed(signClient, trigger.topic)];
                    case 3:
                        session = _d.sent();
                        _d.label = 4;
                    case 4:
                        if (!this.activeAccount) {
                            fun = this.eventHandlers.get("RESET_STATE" /* ClientEvents.RESET_STATE */);
                            fun && fun();
                        }
                        if (!session) {
                            return [2 /*return*/];
                        }
                        _a = this.notifyListeners;
                        _b = [session.pairingTopic];
                        _c = {};
                        return [4 /*yield*/, (0, beacon_utils_1.generateGUID)()];
                    case 5:
                        _a.apply(this, _b.concat([(_c.id = _d.sent(),
                                _c.type = beacon_types_1.BeaconMessageType.Disconnect,
                                _c)]));
                        this.clearState();
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.onPairingClosed = function (signClient, pairingTopic) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var session, error_4;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        session = ((_a = this.session) === null || _a === void 0 ? void 0 : _a.pairingTopic) === pairingTopic
                            ? this.session
                            : signClient.session
                                .getAll()
                                .find(function (session) { return session.pairingTopic === pairingTopic; });
                        if (!session) {
                            return [2 /*return*/, undefined];
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, signClient.disconnect({
                                topic: session.topic,
                                reason: {
                                    code: -1,
                                    message: 'Pairing deleted'
                                }
                            })];
                    case 2:
                        _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        error_4 = _b.sent();
                        // If the session was already closed, `disconnect` will throw an error.
                        console.warn(error_4);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, session];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.onSessionClosed = function (signClient, sessionTopic) {
        return __awaiter(this, void 0, void 0, function () {
            var error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.session || this.session.topic !== sessionTopic) {
                            return [2 /*return*/, undefined];
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, signClient.core.pairing.disconnect({ topic: this.session.pairingTopic })];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        error_5 = _a.sent();
                        // If the pairing was already closed, `disconnect` will throw an error.
                        console.warn(error_5);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, this.session];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.getPairingRequestInfo = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var _b, uri, topic;
            var _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, this.init(true)];
                    case 1:
                        _b = (_a = (_d.sent())) !== null && _a !== void 0 ? _a : {}, uri = _b.uri, topic = _b.topic;
                        _c = {
                            id: topic,
                            type: 'walletconnect-pairing-request',
                            name: 'WalletConnect',
                            version: beacon_core_1.BEACON_VERSION,
                            uri: uri
                        };
                        return [4 /*yield*/, (0, beacon_utils_1.generateGUID)()];
                    case 2:
                        _c.senderId = _d.sent();
                        return [4 /*yield*/, (0, beacon_utils_1.generateGUID)()];
                    case 3: return [2 /*return*/, (_c.publicKey = _d.sent(),
                            _c)];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.closePairings = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var signClient, pairings, _i, pairings_1, pairing;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.closeSessions()];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, this.getSignClient()];
                    case 2:
                        signClient = _b.sent();
                        pairings = (_a = signClient.pairing.getAll()) !== null && _a !== void 0 ? _a : [];
                        _i = 0, pairings_1 = pairings;
                        _b.label = 3;
                    case 3:
                        if (!(_i < pairings_1.length)) return [3 /*break*/, 6];
                        pairing = pairings_1[_i];
                        return [4 /*yield*/, signClient.core.pairing.disconnect({ topic: pairing.topic })];
                    case 4:
                        _b.sent();
                        _b.label = 5;
                    case 5:
                        _i++;
                        return [3 /*break*/, 3];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.closeSessions = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var signClient, sessions, _i, sessions_1, session;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getSignClient()];
                    case 1:
                        signClient = _b.sent();
                        sessions = (_a = signClient.session.getAll()) !== null && _a !== void 0 ? _a : [];
                        _i = 0, sessions_1 = sessions;
                        _b.label = 2;
                    case 2:
                        if (!(_i < sessions_1.length)) return [3 /*break*/, 5];
                        session = sessions_1[_i];
                        return [4 /*yield*/, signClient.disconnect({
                                topic: session.topic,
                                reason: {
                                    code: 0,
                                    message: 'Force new connection'
                                }
                            })];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4:
                        _i++;
                        return [3 /*break*/, 2];
                    case 5:
                        this.clearState();
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.openSession = function (pairingTopic) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function () {
            var signClient, permissionScopeParams, optionalPermissionScopeParams, connectParams, approval, session, error_6, _pairingTopic, errorResponse;
            var _d, _e;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0: return [4 /*yield*/, this.getSignClient()];
                    case 1:
                        signClient = _f.sent();
                        permissionScopeParams = {
                            networks: [this.wcOptions.network],
                            events: [],
                            methods: [
                                PermissionScopeMethods.GET_ACCOUNTS,
                                PermissionScopeMethods.OPERATION_REQUEST,
                                PermissionScopeMethods.SIGN
                            ]
                        };
                        optionalPermissionScopeParams = {
                            networks: [this.wcOptions.network],
                            events: [PermissionScopeEvents.REQUEST_ACKNOWLEDGED],
                            methods: []
                        };
                        connectParams = {
                            requiredNamespaces: (_d = {},
                                _d[TEZOS_PLACEHOLDER] = this.permissionScopeParamsToNamespaces(permissionScopeParams),
                                _d),
                            optionalNamespaces: (_e = {},
                                _e[TEZOS_PLACEHOLDER] = this.permissionScopeParamsToNamespaces(optionalPermissionScopeParams),
                                _e),
                            pairingTopic: pairingTopic !== null && pairingTopic !== void 0 ? pairingTopic : (_a = signClient.core.pairing.getPairings()[0]) === null || _a === void 0 ? void 0 : _a.topic
                        };
                        return [4 /*yield*/, signClient.connect(connectParams)];
                    case 2:
                        approval = (_f.sent()).approval;
                        _f.label = 3;
                    case 3:
                        _f.trys.push([3, 7, , 8]);
                        return [4 /*yield*/, approval()
                            // if I have successfully opened a session and I already have one opened
                        ];
                    case 4:
                        session = _f.sent();
                        if (!(session && this.session)) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.closeSessions()]; // close the previous session
                    case 5:
                        _f.sent(); // close the previous session
                        _f.label = 6;
                    case 6:
                        // I still need this check in the event the user aborts the sync process on the wallet side
                        // but there is already a connection set
                        this.session = (_b = this.session) !== null && _b !== void 0 ? _b : session;
                        this.validateReceivedNamespace(permissionScopeParams, this.session.namespaces);
                        return [3 /*break*/, 8];
                    case 7:
                        error_6 = _f.sent();
                        console.error(error_6.message);
                        _pairingTopic = pairingTopic !== null && pairingTopic !== void 0 ? pairingTopic : (_c = signClient.core.pairing.getPairings()[0]) === null || _c === void 0 ? void 0 : _c.topic;
                        errorResponse = {
                            type: beacon_types_1.BeaconMessageType.Error,
                            id: this.currentMessageId,
                            errorType: beacon_types_1.BeaconErrorType.ABORTED_ERROR
                        };
                        this.notifyListeners(_pairingTopic, errorResponse);
                        return [3 /*break*/, 8];
                    case 8:
                        if (this.session) {
                            return [2 /*return*/, this.session];
                        }
                        else {
                            throw new error_1.InvalidSession('No session set.');
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.permissionScopeParamsToNamespaces = function (permissionScopeParams) {
        var _a;
        return {
            chains: permissionScopeParams.networks.map(function (network) { return "".concat(TEZOS_PLACEHOLDER, ":").concat(network); }),
            methods: permissionScopeParams.methods,
            events: (_a = permissionScopeParams.events) !== null && _a !== void 0 ? _a : []
        };
    };
    WalletConnectCommunicationClient.prototype.validateReceivedNamespace = function (scope, receivedNamespaces) {
        if (receivedNamespaces[TEZOS_PLACEHOLDER]) {
            this.validateMethods(scope.methods, receivedNamespaces[TEZOS_PLACEHOLDER].methods);
            if (scope.events) {
                this.validateEvents(scope.events, receivedNamespaces['tezos'].events);
            }
            this.validateAccounts(scope.networks, receivedNamespaces[TEZOS_PLACEHOLDER].accounts);
        }
        else {
            this.clearState();
            throw new error_1.InvalidReceivedSessionNamespace('All namespaces must be approved', (0, utils_1.getSdkError)('USER_REJECTED').code, 'incomplete', 'tezos');
        }
    };
    WalletConnectCommunicationClient.prototype.validateMethods = function (requiredMethods, receivedMethods) {
        var missingMethods = [];
        requiredMethods.forEach(function (method) {
            if (!receivedMethods.includes(method)) {
                missingMethods.push(method);
            }
        });
        if (missingMethods.length > 0) {
            this.clearState();
            throw new error_1.InvalidReceivedSessionNamespace('All methods must be approved', (0, utils_1.getSdkError)('USER_REJECTED_METHODS').code, 'incomplete', missingMethods);
        }
    };
    WalletConnectCommunicationClient.prototype.validateEvents = function (requiredEvents, receivedEvents) {
        var missingEvents = [];
        requiredEvents.forEach(function (method) {
            if (!receivedEvents.includes(method)) {
                missingEvents.push(method);
            }
        });
        if (missingEvents.length > 0) {
            this.clearState();
            throw new error_1.InvalidReceivedSessionNamespace('All events must be approved', (0, utils_1.getSdkError)('USER_REJECTED_EVENTS').code, 'incomplete', missingEvents);
        }
    };
    WalletConnectCommunicationClient.prototype.validateAccounts = function (requiredNetwork, receivedAccounts) {
        if (receivedAccounts.length === 0) {
            this.clearState();
            throw new error_1.InvalidReceivedSessionNamespace('Accounts must not be empty', (0, utils_1.getSdkError)('USER_REJECTED_CHAINS').code, 'incomplete');
        }
        var receivedChains = [];
        var invalidChains = [];
        var missingChains = [];
        var invalidChainsNamespace = [];
        receivedAccounts.forEach(function (chain) {
            var accountId = chain.split(':');
            if (accountId.length !== 3) {
                invalidChains.push(chain);
            }
            if (accountId[0] !== TEZOS_PLACEHOLDER) {
                invalidChainsNamespace.push(chain);
            }
            var network = accountId[1];
            if (!receivedChains.includes(network)) {
                receivedChains.push(network);
            }
        });
        if (invalidChains.length > 0) {
            this.clearState();
            throw new error_1.InvalidReceivedSessionNamespace('Accounts must be CAIP-10 compliant', (0, utils_1.getSdkError)('USER_REJECTED_CHAINS').code, 'invalid', invalidChains);
        }
        if (invalidChainsNamespace.length > 0) {
            this.clearState();
            throw new error_1.InvalidReceivedSessionNamespace('Accounts must be defined in matching namespace', (0, utils_1.getSdkError)('UNSUPPORTED_ACCOUNTS').code, 'invalid', invalidChainsNamespace);
        }
        requiredNetwork.forEach(function (network) {
            if (!receivedChains.includes(network)) {
                missingChains.push(network);
            }
        });
        if (missingChains.length > 0) {
            this.clearState();
            throw new error_1.InvalidReceivedSessionNamespace('All chains must have at least one account', (0, utils_1.getSdkError)('USER_REJECTED_CHAINS').code, 'incomplete', missingChains);
        }
    };
    WalletConnectCommunicationClient.prototype.validateNetworkAndAccount = function (network, account) {
        if (!this.getTezosNamespace().accounts.includes("".concat(TEZOS_PLACEHOLDER, ":").concat(network, ":").concat(account))) {
            throw new error_1.InvalidNetworkOrAccount(network, account);
        }
    };
    /**
     * @description Access the active network
     * @error ActiveNetworkUnspecified thorwn when there are multiple Tezos netwroks in the session and none is set as the active one
     */
    WalletConnectCommunicationClient.prototype.getActiveNetwork = function () {
        if (!this.activeNetwork) {
            this.getSession();
            throw new error_1.ActiveNetworkUnspecified();
        }
        return this.activeNetwork;
    };
    WalletConnectCommunicationClient.prototype.setDefaultAccountAndNetwork = function () {
        var activeAccount = this.getAccounts();
        if (activeAccount.length === 1) {
            this.activeAccount = activeAccount[0];
        }
        var activeNetwork = this.getNetworks();
        if (activeNetwork.length === 1) {
            this.activeNetwork = activeNetwork[0];
        }
    };
    /**
     * @description Return all connected accounts from the active session
     * @error NotConnected if no active session
     */
    WalletConnectCommunicationClient.prototype.getAccounts = function () {
        return this.getTezosNamespace().accounts.map(function (account) { return account.split(':')[2]; });
    };
    /**
     * @description Return all networks from the namespace of the active session
     * @error NotConnected if no active session
     */
    WalletConnectCommunicationClient.prototype.getNetworks = function () {
        return this.getPermittedNetwork();
    };
    WalletConnectCommunicationClient.prototype.getTezosNamespace = function (namespaces) {
        if (namespaces === void 0) { namespaces = this.getSession().namespaces; }
        if (TEZOS_PLACEHOLDER in namespaces) {
            return namespaces[TEZOS_PLACEHOLDER];
        }
        else {
            throw new error_1.InvalidSession('Tezos not found in namespaces');
        }
    };
    WalletConnectCommunicationClient.prototype.getPermittedMethods = function () {
        return this.getTezosRequiredNamespace().methods;
    };
    WalletConnectCommunicationClient.prototype.getPermittedNetwork = function () {
        return this.getTezosRequiredNamespace().chains.map(function (chain) { return chain.split(':')[1]; });
    };
    WalletConnectCommunicationClient.prototype.getTezosRequiredNamespace = function () {
        return {
            chains: ["".concat(TEZOS_PLACEHOLDER, ":").concat(this.wcOptions.network)],
            events: [],
            methods: ['tezos_getAccounts', 'tezos_send', 'tezos_sign']
        };
        // if (TEZOS_PLACEHOLDER in this.getSession().requiredNamespaces) {
        //   return this.getSession().requiredNamespaces[TEZOS_PLACEHOLDER] as {
        //     chains: string[]
        //     methods: string[]
        //     events: string[]
        //   }
        // } else {
        //   throw new InvalidSession('Tezos not found in requiredNamespaces')
        // }
    };
    WalletConnectCommunicationClient.prototype.notifyListeners = function (pairingTopic, partialResponse) {
        return __awaiter(this, void 0, void 0, function () {
            var response, serializer, serialized;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        response = __assign(__assign({}, partialResponse), { version: '2', senderId: pairingTopic });
                        serializer = new beacon_core_1.Serializer();
                        return [4 /*yield*/, serializer.serialize(response)];
                    case 1:
                        serialized = _a.sent();
                        this.activeListeners.forEach(function (listener) {
                            listener(serialized);
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.currentSession = function () {
        return this.session;
    };
    WalletConnectCommunicationClient.prototype.getSignClient = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(this.signClient === undefined)) return [3 /*break*/, 2];
                        _a = this;
                        return [4 /*yield*/, sign_client_1.SignClient.init(this.wcOptions.opts)];
                    case 1:
                        _a.signClient = _b.sent();
                        this.subscribeToSessionEvents(this.signClient);
                        _b.label = 2;
                    case 2: return [2 /*return*/, this.signClient];
                }
            });
        });
    };
    WalletConnectCommunicationClient.prototype.getSession = function () {
        if (!this.session) {
            throw new error_1.NotConnected();
        }
        return this.session;
    };
    /**
     * @description Access the public key hash of the active account
     * @error ActiveAccountUnspecified thrown when there are multiple Tezos account in the session and none is set as the active one
     */
    WalletConnectCommunicationClient.prototype.getPKH = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!this.activeAccount) {
                    this.getSession();
                    throw new error_1.ActiveAccountUnspecified();
                }
                return [2 /*return*/, this.activeAccount];
            });
        });
    };
    WalletConnectCommunicationClient.prototype.clearState = function () {
        this.session = undefined;
        this.activeAccount = undefined;
        this.activeNetwork = undefined;
    };
    return WalletConnectCommunicationClient;
}(beacon_core_1.CommunicationClient));
exports.WalletConnectCommunicationClient = WalletConnectCommunicationClient;
