// @taquito/taquito/dist/lib/operations/smart-rollup-add-messages-operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartRollupAddMessagesOperation = void 0;
const operations_1 = require("./operations");
/**
 * @description SmartRollupAddMessagesOperation provides utility to fetch properties of SmartRollupAddMessages
 */
class SmartRollupAddMessagesOperation extends operations_1.Operation {
    constructor(hash, params, source, raw, results, context) {
        super(hash, raw, results, context);
        this.params = params;
        this.source = source;
    }
    get operationResults() {
        const smartRollupAddMessagesOp = Array.isArray(this.results) &&
            this.results.find((op) => op.kind === 'smart_rollup_add_messages');
        const result = smartRollupAddMessagesOp &&
            smartRollupAddMessagesOp.metadata &&
            smartRollupAddMessagesOp.metadata.operation_result;
        return result ? result : undefined;
    }
    get status() {
        var _a, _b;
        return (_b = (_a = this.operationResults) === null || _a === void 0 ? void 0 : _a.status) !== null && _b !== void 0 ? _b : 'unknown';
    }
    get message() {
        return this.params.message;
    }
    get fee() {
        return Number(this.params.fee);
    }
    get gasLimit() {
        return Number(this.params.gas_limit);
    }
    get storageLimit() {
        return Number(this.params.storage_limit);
    }
    get consumedMilliGas() {
        var _a;
        return (_a = this.operationResults) === null || _a === void 0 ? void 0 : _a.consumed_milligas;
    }
    get errors() {
        var _a;
        return (_a = this.operationResults) === null || _a === void 0 ? void 0 : _a.errors;
    }
}
exports.SmartRollupAddMessagesOperation = SmartRollupAddMessagesOperation;
