// @taquito/michelson-encoder/dist/lib/tokens/signature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignatureToken = exports.SignatureValidationError = void 0;
const token_1 = require("./token");
const utils_1 = require("@taquito/utils");
/**
 *  @category Error
 *  @description Error that indicates a failure happening when parsing encoding/executing a Signature
 */
class SignatureValidationError extends token_1.TokenValidationError {
    constructor(value, token, message) {
        super(value, token, message);
        this.value = value;
        this.token = token;
        this.name = 'SignatureValidationError';
    }
}
exports.SignatureValidationError = SignatureValidationError;
class SignatureToken extends token_1.ComparableToken {
    constructor(val, idx, fac) {
        super(val, idx, fac);
        this.val = val;
        this.idx = idx;
        this.fac = fac;
    }
    Execute(val) {
        if (val.string) {
            return val.string;
        }
        // TODO decode the signature
        return val.bytes;
    }
    /**
     * @throws {@link SignatureValidationError}
     */
    validate(value) {
        if (utils_1.validateSignature(value) !== utils_1.ValidationResult.VALID) {
            throw new SignatureValidationError(value, this, 'Signature is not valid');
        }
    }
    /**
     * @throws {@link SignatureValidationError}
     */
    Encode(args) {
        const val = args.pop();
        this.validate(val);
        return { string: val };
    }
    /**
     * @throws {@link SignatureValidationError}
     */
    EncodeObject(val, semantic) {
        this.validate(val);
        if (semantic && semantic[SignatureToken.prim]) {
            return semantic[SignatureToken.prim](val);
        }
        return { string: val };
    }
    /**
     * @deprecated ExtractSchema has been deprecated in favor of generateSchema
     *
     */
    ExtractSchema() {
        return SignatureToken.prim;
    }
    generateSchema() {
        return {
            __michelsonType: SignatureToken.prim,
            schema: SignatureToken.prim,
        };
    }
    ToKey(val) {
        return this.Execute(val);
    }
    ToBigMapKey(val) {
        return {
            key: { string: val },
            type: { prim: SignatureToken.prim },
        };
    }
    findAndReturnTokens(tokenToFind, tokens) {
        if (SignatureToken.prim === tokenToFind) {
            tokens.push(this);
        }
        return tokens;
    }
}
exports.SignatureToken = SignatureToken;
SignatureToken.prim = 'signature';
