// @airgap/beacon-types/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorMode = exports.defaultValues = exports.StorageKey = exports.Storage = exports.TransportType = exports.TransportStatus = exports.BeaconErrorType = exports.ExtensionMessageTarget = exports.SigningType = exports.Origin = exports.PermissionScope = exports.BeaconMessageType = exports.NetworkType = exports.TezosOperationType = void 0;
var BeaconMessageType_1 = require("./types/beacon/BeaconMessageType");
Object.defineProperty(exports, "BeaconMessageType", { enumerable: true, get: function () { return BeaconMessageType_1.BeaconMessageType; } });
var PermissionScope_1 = require("./types/beacon/PermissionScope");
Object.defineProperty(exports, "PermissionScope", { enumerable: true, get: function () { return PermissionScope_1.PermissionScope; } });
var NetworkType_1 = require("./types/beacon/NetworkType");
Object.defineProperty(exports, "NetworkType", { enumerable: true, get: function () { return NetworkType_1.NetworkType; } });
var OperationTypes_1 = require("./types/tezos/OperationTypes");
Object.defineProperty(exports, "TezosOperationType", { enumerable: true, get: function () { return OperationTypes_1.TezosOperationType; } });
var Origin_1 = require("./types/Origin");
Object.defineProperty(exports, "Origin", { enumerable: true, get: function () { return Origin_1.Origin; } });
var ExtensionMessageTarget_1 = require("./types/ExtensionMessageTarget");
Object.defineProperty(exports, "ExtensionMessageTarget", { enumerable: true, get: function () { return ExtensionMessageTarget_1.ExtensionMessageTarget; } });
var BeaconErrorType_1 = require("./types/BeaconErrorType");
Object.defineProperty(exports, "BeaconErrorType", { enumerable: true, get: function () { return BeaconErrorType_1.BeaconErrorType; } });
var TransportStatus_1 = require("./types/transport/TransportStatus");
Object.defineProperty(exports, "TransportStatus", { enumerable: true, get: function () { return TransportStatus_1.TransportStatus; } });
var TransportType_1 = require("./types/transport/TransportType");
Object.defineProperty(exports, "TransportType", { enumerable: true, get: function () { return TransportType_1.TransportType; } });
var Storage_1 = require("./types/storage/Storage");
Object.defineProperty(exports, "Storage", { enumerable: true, get: function () { return Storage_1.Storage; } });
var StorageKey_1 = require("./types/storage/StorageKey");
Object.defineProperty(exports, "StorageKey", { enumerable: true, get: function () { return StorageKey_1.StorageKey; } });
var StorageKeyReturnDefaults_1 = require("./types/storage/StorageKeyReturnDefaults");
Object.defineProperty(exports, "defaultValues", { enumerable: true, get: function () { return StorageKeyReturnDefaults_1.defaultValues; } });
var SigningType_1 = require("./types/beacon/SigningType");
Object.defineProperty(exports, "SigningType", { enumerable: true, get: function () { return SigningType_1.SigningType; } });
var ColorMode_1 = require("./types/ColorMode");
Object.defineProperty(exports, "ColorMode", { enumerable: true, get: function () { return ColorMode_1.ColorMode; } });
__exportStar(require("./types/AnalyticsInterface"), exports);
__exportStar(require("./types/beaconV3/PermissionRequest"), exports);
__exportStar(require("./types/ui"), exports);
__exportStar(require("./types/Regions"), exports);
